modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+    :type lhs: `?t`
+
+    :arg rhs: A 
+    :type rhs: `t`
+  */
+  proc moveInitialize(ref lhs: ?t, pragma "no auto destroy" in rhs: t) {
+    if lhs.type != nothing then
+      _move(lhs, rhs);
+  }
+
+  /*
+    Move the memory referred to by a reference into a new value.
+
+    .. warning::
+
+      The memory referred to by ``arg`` should be considered uninitialized
+      after a call to this function.
+
+    :arg arg: A reference to consume
+    :type arg: `?t`
+
+    :return: A reference moved into a new value
+    :rtype: `t`
+  */
+  proc moveToValue(const ref arg: ?t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var result: t;
+      _move(result, arg);
+      return result;
+    }
+  }
+
+  /*
+    Swap the values referred to by two references. No assignment takes
+    place when swapping ``lhs`` and ``rhs``.
+
+    :arg lhs: A value to swap
+    :type lhs: `?t`
+
+    :arg rhs: A value to swap
+    :type rhs: `t`
+  */
+  proc moveSwap(ref lhs: ?t, ref rhs: t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var temp: t;
+      _move(temp, lhs);
+      _move(lhs, rhs);
+      _move(rhs, temp);
+    }
+  }
+
+  private inline proc _haltBadIndex(a, idx, indexName: string) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if !a.domain.contains(idx) then
+      boundsCheckHalt('Cannot move-initialize array because its domain ' +
+                      'does not contain: ' + indexName);
+  }
+
+  // Call to 'indexOrder' assumes 'idx' is a valid index.
+  private inline proc _haltBadElementRange(a, idx, numElements: int) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if numElements > a.size then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy exceeds array size');
+
+    if numElements <= 0 then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy is <= 0');
+
+    const order = a.domain.indexOrder(idx);
+    const hi = order + numElements;
+
+    if hi > a.size then
+      boundsCheckHalt('Cannot move-initialize array because one or ' +
+                      'more indices are outside its domain');
+  }
+
+  private inline proc _errorNonCopyableType(type t) {
+    if !isCopyableType(t) then
+      compilerError('Cannot move-initialize array with non-copyable ',
+                    'element type: ', t:string, 2);
+  }
+
+  // TODO: Relax this restriction in the future.
+  private inline proc _errorNot1DRectangularArray(a) {
+    if !a.isDefaultRectangular() || a.rank > 1 then
+      compilerError('Can only move-initialize one-dimensional ',
+                    'rectangular arrays', 2);
+  }
+
+  /*
+    Move-initialize a group of array elements from a different group of
+    elements in the same array.
```

Comment:
Perhaps add: "This is equivalent to and potentially better-performing than a sequence of individual calls to `moveInitialize().`" Now looking at the code, perhaps "better performing" is better left out.

You need to define what happens when the two sets of indices overlap and present the error conditions that you check for.

Reply:
The reason I omitted those is because it ended up being a ridiculous amount of warnings. Is there a more eloquent way to put it?

Reply:
Possibly:
"This function reports an error when [this normal piece of code] does."
or "Calling this function is legal only when [some piece of code] is."